topic_manager.go

```diff
@@ -105,70 +113,161 @@ func (m *topicManager) Close() error {
 }
 
 func (m *topicManager) Partitions(topic string) ([]int32, error) {
-	return m.client.Partitions(topic)
+	// refresh metadata, otherwise we might get an outdated number of partitions.
+	// we cannot call it for that specific topic,
+	// otherwise we'd create it if auto.create.topics.enable==true, which we want to avoid
+	if err := m.client.RefreshMetadata(); err != nil {
+		return nil, fmt.Errorf("error refreshing metadata %v", err)
+	}
+	topics, err := m.client.Topics()
+	if err != nil {
+		return nil, err
+	}
+	for _, tpc := range topics {
+		// topic exists, let's list the partitions.
+		if tpc == topic {
+			return m.client.Partitions(topic)
+		}
+	}
+	return nil, errTopicNotFound
 }
 
 func (m *topicManager) GetOffset(topic string, partitionID int32, time int64) (int64, error) {
 	return m.client.GetOffset(topic, partitionID, time)
 }
 
-func (m *topicManager) checkTopicExistsWithPartitions(topic string, npar int) (bool, error) {
-	par, err := m.client.Partitions(topic)
-	if err != nil {
-		if err == sarama.ErrUnknownTopicOrPartition {
-			return false, nil
-		}
-		return false, fmt.Errorf("Error checking partitions for topic %s: %v", topic, err)
-	}
-	if len(par) != npar {
-		return false, fmt.Errorf("topic %s has %d partitions instead of %d", topic, len(par), npar)
-	}
-	return true, nil
-}
-
 func (m *topicManager) createTopic(topic string, npar, rfactor int, config map[string]string) error {
+	m.topicManagerConfig.Logger.Debugf("creating topic %s with npar=%d, rfactor=%d, config=%#v", topic, npar, rfactor, config)
 	topicDetail := &sarama.TopicDetail{}
 	topicDetail.NumPartitions = int32(npar)
 	topicDetail.ReplicationFactor = int16(rfactor)
 	topicDetail.ConfigEntries = make(map[string]*string)
 
 	for k, v := range config {
-		topicDetail.ConfigEntries[k] = &v
+		// need to make a copy, or the config values will contain only one value
+		value := v
+		topicDetail.ConfigEntries[k] = &value
 	}
 
-	topicDetails := make(map[string]*sarama.TopicDetail)
-	topicDetails[topic] = topicDetail
+	err := m.admin.CreateTopic(topic, topicDetail, false)
+	if err != nil {
+		return fmt.Errorf("error creating topic %s, npar=%d, rfactor=%d, config=%#v: %v",
+			topic, npar, rfactor, config, err)
+	}
 
-	request := sarama.CreateTopicsRequest{
-		Timeout:      time.Second * 15,
-		TopicDetails: topicDetails,
+	return nil
+}
+
+func (m *topicManager) handleConfigMismatch(message string) error {
+	switch m.topicManagerConfig.MismatchBehavior {
```

Comment:
Do you think this switch is the best option?
If you want a default case I would rather add it explicitly as long every other case also results in a return of nil or an error. Otherwise I would not return in the first case since it's basically just an addition to the default case and thus a redundant code path? But since it's currently equal it does not make a difference.

Reply:
pushed another version. Take a look if that's better. Not entirely sure what you meant here :)